lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -494,36 +499,81 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
     mkLatestGarbageCollectionEpochRef = newIORef minBound
 
     initCursor :: IO [BlockHeader]
-    initCursor = atomically $ readPoolProductionCursor (max 100 k)
+    initCursor = do
+        fromDB <- atomically $ readPoolProductionCursor (max 100 k)
+        pure (lastByronBlock:fromDB)
       where k = fromIntegral $ getQuantity getEpochStability
+            -- We only have stake pools since shelley era.
+            -- This is hardcoded for mainnet and will do nothing for other
+            -- networks.
+            -- TODO: query cardano-node for hard fork block (not yet possible)
+            --       https://jira.iohk.io/browse/CAD-1850
+            lastByronBlock :: BlockHeader
+            lastByronBlock = BlockHeader
+                (SlotNo 4492799)
+                (Quantity 4490510)
+                (Hash $ unsafeFromHex
+                    "f8084c61b6a238acec985b59310b6ecec49c0ab8352249afd7268da5cff2a457")
+                (Hash $ unsafeFromHex
+                    "aa83acbf5904c0edfe4d79b3689d3d00fcfc553cf360fd2229b98d464c28e9de")
 
     getHeader :: CardanoBlock sc -> BlockHeader
     getHeader = toCardanoBlockHeader gp
 
     forward
         :: IORef EpochNo
+        -> IORef Int
         -> NonEmpty (CardanoBlock sc)
         -> (BlockHeader, ProtocolParameters)
         -> IO (FollowAction ())
-    forward latestGarbageCollectionEpochRef blocks (_nodeTip, _pparams) = do
-        atomically $ forM_ blocks $ \case
+    forward latestGarbageCollectionEpochRef chunkCounter blocks (_nodeTip, _pparams) = do
+        atomically $ forML_ blocks forAllBlocks forLastBlock
+        pure Continue
+      where
+        {-# INLINE forAllBlocks #-}
+        forAllBlocks = \case
             BlockByron _ -> pure ()
             BlockShelley blk -> do
                 let (slot, certificates) = poolCertsFromShelleyBlock blk
                 let header = toShelleyBlockHeader getGenesisBlockHash blk
-                runExceptT (putPoolProduction header (getProducer blk))
-                    >>= \case
-                        Left e ->
-                            liftIO $ traceWith tr $ MsgErrProduction e
-                        Right () ->
-                            pure ()
+                handleErr (putPoolProduction header (getProducer blk))
                 unless (null certificates) $ do
                     -- Before adding new pool certificates to the database, we
                     -- first attempt to garbage collect pools that have already
                     -- retired.
                     garbageCollectPools slot latestGarbageCollectionEpochRef
                     putPoolCertificates slot certificates
-        pure Continue
+
+        {-# INLINE forLastBlock #-}
+        forLastBlock = \case
+            BlockByron blk -> do
+                c <- liftIO $ readIORef chunkCounter
+                when (c `mod` 50 == 0) $ do -- every 50k blocks, insert a header
```

Comment:
Is it possible that we might repeatedly skip over this condition? As far as I know, we don't really control how many blocks we're given by the block following protocol. Suppose we're initially given 10 blocks, and then from that point forward, repeatedly given 50 blocks? Would the `c mod 50` condition not then would always evaluate to `10`?

What if instead we just write a header once per epoch? We already track the last seen epoch in the `mkLatestGarbageCollectionEpochRef`. Perhaps there is a way to reuse that reference here, and make it into a more general purpose reference of the "last-seen epoch"?

Whenever we get a new set of blocks, we can check whether or not we've transitioned into a new epoch, and if so:

 1. write a new block header (if we're in the Byron era).
 2. perform garbage collection of retired pools (if we're in the Shelley era).
 3. update the latest epoch reference.

Reply:
What is this even needed? Writing one block header every batch isn't going to be very costly. `forLastBlock` could be simplified to write block headers to the database everytime it is called, there's no need for further filtering.

Reply:
More importantly, the point of writing those headers is to maintain a list of candidates that'll be used for finding an intersection in case the system restart. So, having gaps of 50k blocks (or even, of an entire epoch) is _not acceptable_. 

The chain follower implementation will yield blocks by batches while syncing far from the tip, and once reaching the "unstable edge" (i.e. k blocks from the network tip), it'll start yielding them in a slower fashion. In practice, we want to have some a high density of blocks near the tip, and can resort to a sparse view for deeper blocks. This is implemented by the `sparseCheckpoints` function in `Cardano.Wallet.DB` which gives you a list of block numbers to preserve given `k` and a current chain height

Reply:
The approach taken for wallet workers is to always store the last block received from the chain follower and, regularly _prune_ old checkpoints that are no longer required (because, the chain is always going forward). This works well in practice, so I'll suggest to stick to this :+1:

Reply:
I did all those changes... but note that a regular uninterrupted sync on non-mainnet will be slower due to the inserts.